### PR TITLE
[DEV-1092] metamask foundation

### PR DESCRIPTION
**Pull request by Krysh90**

### CODE REVIEW

File: src/contexts/wallet.context.tsx [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+const WalletContext = createContext<WalletInterface>(undefined as any);
+
+export function useWalletContext(): WalletInterface {
+  return useContext(WalletContext);
+}
+
+export function WalletContextProvider(props: PropsWithChildren): JSX.Element {
+  const { address, setAddress } = useSession();
+  const { getSignMessage } = useApi();
+  const { ethereum } = window as any;
```
Comment (davidleomay) [inline src/contexts/wallet.context.tsx:21]: What is this? 😆

File: src/contexts/wallet.context.tsx [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  const web3 = new Web3(Web3.givenProvider);
+
+  useEffect(() => {
+    web3.eth.getAccounts((err, accounts) => {
+      setAddress(verifyAccount(accounts));
+    });
+    ethereum.on('accountsChanged', (accounts: string[]) => {
+      setAddress(verifyAccount(accounts));
+    });
```
Comment (davidleomay) [inline src/contexts/wallet.context.tsx:30]: What is this doing?

File: src/contexts/wallet.context.tsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+  const verifyAccount = (accounts: string[]): string | undefined => {
+    if (accounts.length <= 0) return undefined;
+    // check if address is valid
+    return Web3.utils.toChecksumAddress(accounts[0]);
+  };
+
+  const isInstalled = (): boolean => {
+    return ethereum && ethereum.isMetaMask;
+  };
```
Comment (davidleomay) [inline src/contexts/wallet.context.tsx:41]: Could just use a property
```ts
const isInstalled = ethereum && ethereum.isMetaMask;
```

**Review by davidleomay — APPROVED** [review]

File: src/contexts/wallet.context.tsx (hunk 1 above)
Comment (Krysh90) [reply src/contexts/wallet.context.tsx:21]: as far as I have understood, metamask browser exchanges makes an ethereum object available on window. This should be used to hook into some functionality. `Web3` additionally is a wrapper, which makes a few things easier, more accessible and typed for better TS support

File: src/contexts/wallet.context.tsx (hunk 2 above)
Comment (Krysh90) [reply src/contexts/wallet.context.tsx:30]: You only need to connect once and you will stay connected with this site forever. The first function is there to read available (connected) accounts. The second function is to add a "listener" to the global ethereum object to receive updates, if the user changes his account in metamask.

**Comment by Krysh90** [comment]
Just rechecked those changes on a browser without MetaMask installed and therefore fixed a few issues, as `ethereum` is not available and received accounts are `undefined`, therefore breaking things.

**Review by davidleomay — APPROVED** [review]

**Pull request merged**